zcash_primitives/src/transaction/txid.rs

```diff
@@ -0,0 +1,580 @@
+use std::borrow::Borrow;
+use std::convert::TryFrom;
+use std::io::Write;
+
+use blake2b_simd::Hash as Blake2bHash;
+use byteorder::{LittleEndian, WriteBytesExt};
+use ff::PrimeField;
+use group::GroupEncoding;
+use orchard::bundle::{self as orchard};
+
+use crate::consensus::{BlockHeight, BranchId};
+
+use super::{
+    blake2b_256::HashWriter,
+    components::{
+        amount::Amount,
+        orchard as ser_orch,
+        sapling::{self, OutputDescription, SpendDescription},
+        transparent::{self, TxIn, TxOut},
+    },
+    Authorization, Authorized, TransactionDigest, TransparentDigests, TxDigests, TxId, TxVersion,
+};
+
+#[cfg(feature = "zfuture")]
+use super::{
+    components::tze::{self, TzeIn, TzeOut},
+    TzeDigests,
+};
+
+/// TxId tree root personalization
+const ZCASH_TX_PERSONALIZATION_PREFIX: &[u8; 12] = b"ZcashTxHash_";
+
+// TxId level 1 node personalization
+const ZCASH_HEADERS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdHeadersHash";
+const ZCASH_TRANSPARENT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdTranspaHash";
+const ZCASH_SAPLING_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSaplingHash";
+const ZCASH_ORCHARD_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdOrchardHash";
+#[cfg(feature = "zfuture")]
+const ZCASH_TZE_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdTZE____Hash";
+
+// TxId transparent level 2 node personalization
+const ZCASH_PREVOUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdPrevoutHash";
+const ZCASH_SEQUENCE_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSequencHash";
+const ZCASH_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdOutputsHash";
+
+// TxId tze level 2 node personalization
+#[cfg(feature = "zfuture")]
+const ZCASH_TZE_INPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdTZEIns_Hash";
+#[cfg(feature = "zfuture")]
+const ZCASH_TZE_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdTZEOutsHash";
+
+// TxId sapling level 2 node personalization
+const ZCASH_SAPLING_SPENDS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSSpendsHash";
+const ZCASH_SAPLING_SPENDS_COMPACT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSSpendCHash";
+const ZCASH_SAPLING_SPENDS_NONCOMPACT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSSpendNHash";
+
+const ZCASH_SAPLING_OUTPUTS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSOutputHash";
+const ZCASH_SAPLING_OUTPUTS_COMPACT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSOutC__Hash";
+const ZCASH_SAPLING_OUTPUTS_MEMOS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSOutM__Hash";
+const ZCASH_SAPLING_OUTPUTS_NONCOMPACT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdSOutN__Hash";
+
+const ZCASH_ORCHARD_ACTIONS_COMPACT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdOrcActCHash";
+const ZCASH_ORCHARD_ACTIONS_MEMOS_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdOrcActMHash";
+const ZCASH_ORCHARD_ACTIONS_NONCOMPACT_HASH_PERSONALIZATION: &[u8; 16] = b"ZTxIdOrcActNHash";
```

Comment:
We should update ZIP 225 (https://zips.z.cash/zip-0225) which still uses `OrcOut` instead of `OrcAct`.

Reply:
It's part of this PR: https://github.com/zcash/zips/pull/505